xds/src/main/java/io/grpc/xds/OrcaUtil.java

```diff
@@ -0,0 +1,203 @@
+    return new OrcaReportingTracerFactory(delegate, listener);
+  }
+
+  /**
+   * Creates a new {@link LoadBalancer.Helper} with provided {@link OrcaReportListener} installed to
+   * receive callback when an out-of-band ORCA report is received.
+   *
+   * @param delegate the delegate helper that provides essentials for establishing subchannels to
+   *     backends.
+   * @param listener contains the callback to be invoked when an out-of-band ORCA report is
+   *     received.
+   * @param maxReportInterval the maximum expected interval of receiving periodical ORCA reports.
+   * @param unit time unit of {@code maxReportInterval} value.
+   */
+  public static LoadBalancer.Helper newOrcaReportingHelper(
+      LoadBalancer.Helper delegate,
+      OrcaReportListener listener,
+      long maxReportInterval,
+      TimeUnit unit) {
+    // TODO(chengyuanzhang): create a LoadReportingHelper that intercepts createSubChannel() method.
+    // This requires OOB ORCA service impl to be done first in order to create an OrcaClient.
+    return null;
+  }
+
+  /**
+   * The listener interface for receiving backend ORCA reports. The class that is interested in
+   * processing backend cost metrics implements this interface, and the object created with that
+   * class is registered with a component, using methods in {@link OrcaUtil}. When an ORCA report is
+   * received, that object's {@code onLoadReport} method is invoked.
+   */
+  public interface OrcaReportListener {
+
+    /**
+     * Invoked when an ORCA report is received.
+     */
+    void onLoadReport(OrcaLoadReport report);
+  }
+
+  @VisibleForTesting
+  static final class OrcaReportingTracerFactory extends ClientStreamTracer.Factory {
+
+    @VisibleForTesting
+    static final CallOptions.Key<OrcaReportBroker> ORCA_REPORT_BROKER_KEY =
+        CallOptions.Key.create("internal-orca-report-broker");
+    @VisibleForTesting
+    static final Metadata.Key<OrcaLoadReport> ORCA_ENDPOINT_LOAD_METRICS_KEY =
+        Metadata.Key.of(
+            "x-endpoint-load-metrics-bin",
+            ProtoUtils.metadataMarshaller(OrcaLoadReport.getDefaultInstance()));
+
+    private final ClientStreamTracer.Factory delegate;
+    private final OrcaReportListener listener;
+
+    OrcaReportingTracerFactory(
+        ClientStreamTracer.Factory delegate, OrcaReportListener listener) {
+      this.delegate = checkNotNull(delegate, "delegate");
+      this.listener = checkNotNull(listener, "listener");
+    }
+
+    @Override
+    public ClientStreamTracer newClientStreamTracer(StreamInfo info, Metadata headers) {
+      OrcaReportBroker broker = info.getCallOptions().getOption(ORCA_REPORT_BROKER_KEY);
+      boolean augmented = false;
+      if (broker == null) {
+        broker = new OrcaReportBroker();
+        final StreamInfo currInfo = info;
+        final OrcaReportBroker currBroker = broker;
+        info = new ClientStreamTracer.StreamInfo() {
```

Comment:
@ejona86 I think `StreamInfo` should've been a final class with builder rather than an abstract class.
The introduction of `ForwardingClientStreamTracer` endorses the act of wrapping, there may be a need for altering parts of `StreamInfo` before passing it to the delegate. A `toBuilder()` or a `ForwardingStreamInfo` will be needed, and the former sounds better.

Reply:
@zhangkun83 I agree it should be something that `CreateSubchannelArgs` is doing.

Reply:
final class + builder SGTM

Reply:
Sent #5648